### PR TITLE
cpu/stm32: add support for U5 family

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds support for the new [STM32 U5 series](https://www.st.com/en/microcontrollers-microprocessors/stm32u5-series.html) and adds support for the [B-U585i-IOT02A](https://www.st.com/en/evaluation-tools/b-u585i-iot02a.html#overview) development kit board.

The U5 is am M33, similar to the L5 but with more RAM/Flash. As usual, there are quite some differences, especially with the clock management (2 MSIs, registers to configure the PLL are different, etc).
The port is very basic: only GPIO, UART, Flashpage and Timer peripherals are tested. There's also a configuration for I2C (that should work) but is not tested yet.
SPI registers have changed names so the driver needs to be adapted.
There's no clock configuration via Kconfig available. I don't know if that's mandatory or not.

### Testing procedure

- Green CI
- Run `compile_and_test_for_board.py`:

```
ERROR:b-u585i-iot02a:Tests failed: 6
Failures during compilation:
- [tests/pkg_cifra](tests/pkg_cifra/compilation.failed)

Failures during test:
- [tests/malloc](tests/malloc/test.failed)
- [tests/malloc_thread_safety](tests/malloc_thread_safety/test.failed)
- [tests/periph_timer_short_relative_set](tests/periph_timer_short_relative_set/test.failed)
- [tests/pthread_rwlock](tests/pthread_rwlock/test.failed)
- [tests/sys_sema_inv](tests/sys_sema_inv/test.failed)
```

Some failures are related to RAM. I don't know yet what's the problem with `tests/pkg_cifra`.

### Issues/PRs references

None

### CODE REVIEW

File: cpu/stm32/stmclk/stmclk_u5.c [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+#error "PLL configuration: PLL R value is out of range"
+#endif
+#define PLL_R                       ((CONFIG_CLOCK_PLL_R - 1) << RCC_PLL1DIVR_PLL1R_Pos)
+
+#if (CONFIG_CLOCK_PLL_Q < 1 || CONFIG_CLOCK_PLL_Q > 128)
+#error "PLL configuration: PLL Q value is out of range"
+#endif
+#define PLL_Q                       ((CONFIG_CLOCK_PLL_Q - 1) << RCC_PLL1DIVR_PLL1Q_Pos)
+
+/* Define MSI range bitfields */
```
Comment (aabadie) [inline cpu/stm32/stmclk/stmclk_u5.c:67]: TODO: fix the range defines (only 48MHz is correct)

**Comment by benpicco** [comment]
Code looks good!
What is the output of `tests/malloc`?

**Comment by aabadie** [comment]
> What is the output of tests/malloc?

I just keep the end (and after increasing the test timeout to 30s):

```
Allocated 128 Bytes at 0x0x200bf880, total 697536
Allocated 128 Bytes at 0x0x200bf918, total 697672
Allocated 128 Bytes at 0x0x200bf9b0, total 697808
Allocated 128 Bytes at 0x0x200bfa48, total 697944
Allocated 128 Bytes at 0x0x200bfae0, total 698080
Allocated 128 Bytes at 0x0x200bfb78, total 698216
Allocated 128 Bytes at 0x0x200bfc10, total 698352
Allocated 128 Bytes at 0x0x200bfca8, total 698488
Allocated 128 Bytes at 0x0x200bfd40, total 698624
Allocated 128 Bytes at 0x0x200bfdd8, total 698760
Allocated 128 Bytes at 0x0x200bfe70, total 698896
Allocated 128 Bytes at 0x0x200bff08, total 699032

Context before hardfault:
   r0: 0xffffffff
   r1: 0x200c0020
   r2: 0x0000000c
   r3: 0x00000000
  r12: 0x080023ef
   lr: 0x080014c7
   pc: 0x08001536
  psr: 0x61000000

FSR/FAR:
 CFSR: 0x00008200
 HFSR: 0x40000000
 DFSR: 0x0000000b
 AFSR: 0x00000000
 BFAR: 0x00000000
Misc
EXC_RET: 0xffffffbc
Active thread: 1 "main"
Attempting to reconstruct state for debugging...
In GDB:
  set $pc=0x8001536
  frame 0
  bt

ISR stack overflowed by at least 16 bytes.
^C
```

GDB doesn't give much information. I tried to change the RAM_LEN to something smaller (that corresponds to SRAM1/2/3 or just SRAM1) but it crashes the same. I don't understand the problem.

For the record, here are the outputs of the other failing tests:

<details><summary>tests/malloc_thread_safety</summary>

```
START
main(): This is RIOT! (Version: 2022.01-devel-1317-g5674d-pr/cpu/stm32u5)
Test Application for multithreaded use of malloc()
==================================================

This test will run duelling threads allocating and freeing memory.
The running thread is interrupted every millisecond and the other
threads gets scheduled. Eventually, this should yield to memory
corruption unless proper guards are in place preventing them. After
ca. two seconds without crash, the test is considered as passing.

Testing: malloc()/free()
Stack pointer corrupted, reset to top of stack
FSR/FAR:
 CFSR: 0x00008200
 HFSR: 0x40000000
 DFSR: 0x0000000b
 AFSR: 0x00000000
 BFAR: 0x00000004
Misc
EXC_RET: 0xffffffb0
Timeout in expect script at "child.expect("TEST PASSED")" (tests/malloc_thread_safety/tests/01-run.py:16)


```

</details>

<details><summary>tests/pthread_rwlock</summary>

```
main(): This is RIOT! (Version: 2022.01-devel-1317-g5674d-pr/cpu/stm32u5)
START
start
d2 (prio=8): sleep for    69522 µs.
start
d3 (prio=8): sleep for    95682 µs.
start
start
start
start
start
start
r2 (prio=8): 0: read  <-  0 (correct = 1)
d2 (prio=8): sleep for    80836 µs.
r3 (prio=8): 0: read  <-  0 (correct = 1)
d3 (prio=8): sleep for    20228 µs.
d7 (prio=8): sleep for   279012 µs.
w7 (prio=8): 0: write ->  1 (correct = 1)
d7 (prio=8): sleep for    42710 µs.
d3 (prio=8): sleep for    72654 µs.
d2 (prio=8): sleep for    52937 µs.
d4 (prio=9): sleep for    28038 µs.
d5 (prio=9): sleep for    25251 µs.
d6 (prio=9): sleep for    20994 µs.
r6 (prio=9): 0: read  <-  1 (correct = 1)
d6 (prio=9): sleep for    97095 µs.
r4 (prio=9): 0: read  <-  1 (correct = 1)
d4 (prio=9): sleep for    18099 µs.
r5 (prio=9): 0: read  <-  1 (correct = 1)
d5 (prio=9): sleep for     7884 µs.
r2 (prio=8): 1: read  <-  1 (correct = 1)
d2 (prio=8): sleep for    25930 µs.
r3 (prio=8): 1: read  <-  1 (correct = 1)
d3 (prio=8): sleep for    43392 µs.
d7 (prio=8): sleep for   278259 µs.
w7 (prio=8): 1: write ->  2 (correct = 1)
d7 (prio=8): sleep for   123248 µs.
d2 (prio=8): sleep for    55317 µs.
d3 (prio=8): sleep for    46098 µs.
r3 (prio=8): 2: read  <-  2 (correct = 1)
d3 (prio=8): sleep for    78386 µs.
r2 (prio=8): 2: read  <-  2 (correct = 1)
d2 (prio=8): sleep for    55048 µs.
d8 (prio=9): sleep for   167181 µs.
w8 (prio=9): 0: write ->  3 (correct = 1)
d2 (prio=8): sleep for    23545 µs.
d8 (prio=9): sleep for    92530 µs.
r2 (prio=8): 3: read  <-  3 (correct = 1)
d2 (prio=8): sleep for    88291 µs.
d7 (prio=8): sleep for   109749 µs.
w7 (prio=8): 2: write ->  4 (correct = 1)
d7 (prio=8): sleep for   194090 µs.
d3 (prio=8): sleep for    87987 µs.
d2 (prio=8): sleep for     5106 µs.
r2 (prio=8): 4: read  <-  4 (correct = 1)
d2 (prio=8): sleep for    38803 µs.
done
Stack pointer corrupted, reset to top of stack
FSR/FAR:
 CFSR: 0x00008200
 HFSR: 0x40000000
 DFSR: 0x0000000b
 AFSR: 0x00000000
 BFAR: 0x00000004
Misc
EXC_RET: 0xffffffb0
Timeout in expect script at "child.expect('done')" (tests/pthread_rwlock/tests/01-run.py:13)
```

</details>


<details><summary>tests/sys_sema_inv</summary>

```
START
main(): This is RIOT! (Version: 2022.01-devel-1317-g5674d-pr/cpu/stm32u5)
counter mode
THREAD 1 start
THREAD 2 start
THREAD 3 start
Stack pointer corrupted, reset to top of stack
FSR/FAR:
 CFSR: 0x00008200
 HFSR: 0x40000000
 DFSR: 0x0000000b
 AFSR: 0x00000000
 BFAR: 0x00000004
Misc
EXC_RET: 0xffffffb0
Timeout in expect script at "child.expect_exact('thread synced')" (tests/sys_sema_inv/tests/01-run.py:12)
```

</details>

**Comment by aabadie** [comment]
> "SRAM1/2/3 Non-secure" only goes to 0x200C0000 so that explains the crash in the malloc test. There is a gap and SRAM4 starts at 0x28000000.

Yes, I saw that gap, that's why I only set RAM_LEN to be SRAM1 + 2 + 3 (768K) but it still crashes. Even with super small values of RAM_LEN (like 64K).

**Comment by aabadie** [comment]
Just tried the first I2C configured with a MEMs shield, it works:

```
2021-12-17 11:08:13,889 # main(): This is RIOT! (Version: 2022.01-devel-1319-g37856-pr/cpu/stm32u5)
2021-12-17 11:08:13,891 # Init HTS221 on I2C_DEV(0)
2021-12-17 11:08:13,903 # H: 72.6%, T: 20.9°C
2021-12-17 11:08:15,906 # H: 72.4%, T: 21.0°C
```

**Comment by benpicco** [comment]
> but it still crashes. Even with super small values of RAM_LEN (like 64K).

huh that's strange - what address is it crashing on then?

**Comment by aabadie** [comment]
> what address is it crashing on then?

When I set 64K, it's crashing at `r1: 0x20010008`, with 768K, it crashes at `r1: 0x200c0020` (not sure if r1 is the right value to check but it seems to be that).

Maybe there's a reserved space ? I couldn't find any information about that.

**Comment by aabadie** [comment]
There's an issue with Murdock (different between TEST_KCONFIG=0 and TEST_KCONFIG=1) but I don't know how to fix it. I can also see locally the 16B difference between both builds.

**Comment by aabadie** [comment]
I could fix the firmware difference between Kconfig and no Kconfig. It was related to the CLOCK_APB1/APB2 constants being different.

Regarding the issues with the tests, the problem is general to Cortex-M33 and the ARM toolchain version, the same tests are failing on nucleo-l552ze-q. For `tests/malloc` I could find a way to make it work: use version 2021.10 of the Docker image or ARM GCC 9. With ARM GCC 10, it doesn't work. For the other tests, they apparently never worked but I find this strange because in #15192, `tests/pthread_rwlock` was not reported in the test failures.

**Comment by benpicco** [comment]
Do we have those problems on Cortex-M33 documented somewhere? 
That sounds like a serious issue. @dylad did you also see this on nRF9160?

File: boards/b-u585i-iot02a/doc.txt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+and debug via GDB by simply typing
+```
+make BOARD=b-u585i-iot02a debug
+```
+
+## Supported Toolchains
+
+For using the ST B-U585I-IOT02A board we recommend the usage of the
+[GNU Tools for ARM Embedded Processors](https://launchpad.net/gcc-arm-embedded)
+toolchain.
```
Comment (benpicco) [inline boards/b-u585i-iot02a/doc.txt:38]: Better add a warning about the issue you've experienced here.

**Comment by benpicco** [comment]
Does this work as `ldscripts/stm32u5.ld` with `LINKER_SCRIPT ?= stm32u5.ld` and `NUM_HEAPS` set to 2?

```
INCLUDE cortexm.ld

MEMORY
{
    sram4       : ORIGIN = 0x28000000, LENGTH = 16K
}

SECTIONS 
{
    .heap2 ALIGN(4) (NOLOAD) :
    {
        _sheap2 = . ;
        _eheap2 = ORIGIN(sram4) + LENGTH(sram4);
    } > sram4  
}

```

**Comment by aabadie** [comment]
> Do we have those problems on Cortex-M33 documented somewhere?

No and I think I'll open an issue about them.

**Comment by dylad** [comment]
> That sounds like a serious issue. @dylad did you also see this on nRF9160?

I didn't notice any issue on nRF9160. I'll try to re-run `tests/malloc` on it this weekend and report back if I spot something strange.

**Comment by aabadie** [comment]
> Does this work as ldscripts/stm32u5.ld with LINKER_SCRIPT ?= stm32u5.ld and NUM_HEAPS set to 2?

I get the same failures but it doesn't seem to cause other troubles.

**Comment by aabadie** [comment]
> I think I'll open an issue about them.

Done => #17439

I also added 7e3256a with your suggestion about the second heap.

File: boards/b-u585i-iot02a/doc.txt (hunk 2 above)
Comment (aabadie) [reply boards/b-u585i-iot02a/doc.txt:38]: Added a note about that.

File: cpu/stm32/Makefile.include [hunk 3]
```diff
@@ -1,7 +1,12 @@
 CFLAGS += -DCPU_FAM_STM32$(call uppercase_and_underscore,$(CPU_FAM))
 
-# For stm32 cpu's we use the generic stm32.ld linker script
-LINKER_SCRIPT ?= stm32.ld
+# For stm32 cpu's we use the generic stm32.ld linker script,
+# stm32u5 has it's own linker script
+ifeq (u5,$(CPU_FAM))
+  LINKER_SCRIPT ?= stm32u5.ld
+else
+  LINKER_SCRIPT ?= stm32.ld
```
Comment (benpicco) [inline cpu/stm32/Makefile.include:8]: Oh I didn't realize we already have a custom linker script for stm32 - then let's use this instead

```
diff --git a/cpu/stm32/Makefile.include b/cpu/stm32/Makefile.include
index 884890ccc1..406f58a34c 100644
--- a/cpu/stm32/Makefile.include
+++ b/cpu/stm32/Makefile.include
@@ -61,6 +61,9 @@ info-stm32:
 ifneq (,$(CCMRAM_LEN))
   LINKFLAGS += $(LINKFLAGPREFIX)--defsym=_ccmram_length=$(CCMRAM_LEN)
 endif
+ifneq (,$(SRAM4_LEN))
+  LINKFLAGS += $(LINKFLAGPREFIX)--defsym=_sram4_length=$(SRAM4_LEN)
+endif
 
 VECTORS_O ?= $(BINDIR)/stm32_vectors/$(CPU_LINE).o
 VECTORS_FILE = $(RIOTCPU)/stm32/vectors/$(CPU_LINE).c
diff --git a/cpu/stm32/ldscripts/stm32.ld b/cpu/stm32/ldscripts/stm32.ld
index fb50fcf119..3e429d879d 100644
--- a/cpu/stm32/ldscripts/stm32.ld
+++ b/cpu/stm32/ldscripts/stm32.ld
@@ -19,10 +19,12 @@
  */
 
 ccmram_length = DEFINED( ccmram_len ) ? ccmram_len : 0x0 ;
+sram4_length = DEFINED( sram4_length ) ? sram4_length : 0x0 ;
 
 MEMORY
 {
     ccmram  : ORIGIN = 0x10000000, LENGTH = ccmram_length
+    sram4   : ORIGIN = 0x28000000, LENGTH = sram4_length
 }
 
 INCLUDE cortexm.ld
```

**Comment by dylad** [comment]
nRF9160 seems not affect by your issue. `tests/malloc` runs without any problem on it.

**Comment by aabadie** [comment]
> nRF9160 seems not affect by your issue. tests/malloc runs without any problem on it.

Are you using ARM GCC 10 ?

**Comment by dylad** [comment]
> Are you using ARM GCC 10 ?
Yes, my local version is `gcc version 10.2.1 20201103 (release) (GNU Arm Embedded Toolchain 10-2020-q4-major) `

File: cpu/stm32/Makefile.include (hunk 3 above)
Comment (benpicco) [reply cpu/stm32/Makefile.include]: Ah but why remove

```C
#define NUM_HEAPS   2
```
from `cpu_conf.h `?

Comment (aabadie) [reply cpu/stm32/Makefile.include]: I don't really know why it's required, I thought it was useless.

File: cpu/stm32/ldscripts/stm32.ld [hunk 4]
```diff
@@ -19,10 +19,12 @@
  */
 
 ccmram_length = DEFINED( ccmram_len ) ? ccmram_len : 0x0 ;
+sram4_length = DEFINED( sram4_length ) ? sram4_length : 0x0 ;
 
 MEMORY
 {
     ccmram  : ORIGIN = 0x10000000, LENGTH = ccmram_length
+    sram4   : ORIGIN = 0x28000000, LENGTH = sram4_length
 }
```
Comment (benpicco) [inline cpu/stm32/ldscripts/stm32.ld:29]:
```suggestion

    .heap2 ALIGN(4) (NOLOAD) :
    {
        _sheap2 = . ;
        _eheap2 = ORIGIN(sram4) + LENGTH(sram4);
    } > sram4
```

File: cpu/stm32/Makefile.include (hunk 3 above)
Comment (benpicco) [reply cpu/stm32/Makefile.include]: It's used by [`syscalls.c`](https://github.com/RIOT-OS/RIOT/blob/master/sys/newlib_syscalls_default/syscalls.c#L106) to provide non-continuous memory sections as additional heap used by `malloc`

**Comment by benpicco** [comment]
Oh but but `NUM_HEAPS` = 2 corresponds to `_sheap1` being present - this is typically [backup RAM](https://github.com/RIOT-OS/RIOT/blob/master/cpu/cortexm_common/ldscripts/cortexm_base.ld#L232). If STM32U5 also has that, we can bump `NUM_HEAPS` to 3 and all is fine. But this would require #16870.

So for now we can change SRAM4 to `_sheap1` and change it back to `_sheap2` when adding backup RAM support.

**Comment by aabadie** [comment]
> So for now we can change SRAM4 to _sheap1 and change it back to _sheap2 when adding backup RAM support.

I think I've done that but not sure.

**Review by benpicco — APPROVED** [review]

**Comment by aabadie** [comment]
Thanks @benpicco for reviewing and merging !

**Comment by gschorcht** [comment]
@aabadie @benpicco 

> > What is the output of tests/malloc?
> 
> I just keep the end (and after increasing the test timeout to 30s):
> 
> ```
> ...
> Allocated 1024 Bytes at 0x200bf568, total 767800
> Allocated 1024 Bytes at 0x200bf980, total 768832
> 
> Context before hardfault:
>    r0: 0xffffffff
>    r1: 0x200c0198
>    r2: 0x0000000c
>    r3: 0x00000000
>   r12: 0x0000000a
>    lr: 0x0800167b
>    pc: 0x080016ea
>   psr: 0x61000000
> ```

I tried to figure out what the problem is. The hard fault happens here (debugger output)
```python
  r0   0xffffffff  -1              r1    0x200c0198   537657752
  r2   0xc         12              r3    0x0          0
  r4   0x0         0               r5    0x408        1032
  r6   0x0         0               r7    0x20000220   536871456
  xpsr 0x61000000  1627389952      fpscr 0x0          0
  msp  0x200001c8  0x200001c8      psp   0x20000800   0x20000800
...
     0x80016d8 <_malloc_r+56>        bl      0x8001660 <sbrk_aligned>
     0x80016dc <_malloc_r+60>        adds    r3, r0, #1
     0x80016de <_malloc_r+62>        mov     r4, r0
     0x80016e0 <_malloc_r+64>        bne.n   0x800177e <_malloc_r+222>
     0x80016e2 <_malloc_r+66>        ldr     r4, [r6, #0]
     0x80016e4 <_malloc_r+68>        mov     r6, r4
     0x80016e6 <_malloc_r+70>        cmp     r6, #0
     0x80016e8 <_malloc_r+72>        bne.n   0x800176c <_malloc_r+204>
  >  0x80016ea <_malloc_r+74>        ldr     r3, [r4, #0]
```
after calling `sbrk_aligned` when it returns -1 (`0xffffffff`) in `r0`.

The return value from `sbrk_aligned` is tested for -1 and a jump to `_malloc_r+222` is executed @`0x80016e0` if it is not -1.

However, if `sbrk_aligned` returns -1, `r4` is loaded with `[r6 + 0x0]` @`0x80016e2` that is just the `__malloc_free_list` which is `NULL` at that moment. After copying `r4` to `r6` @`0x80016e4`, `r4` and `r6` then contain the value of `__malloc_free_list`, that is `0x0` at the moment.

So far so good. The comparison of `r6` with `0x0` @`0x80016e6` and the following conditional jump should prevent to execute the instruction @0x80016ea with `r4` == 0x0, but this doesn't seem to work :thinking: although the Z-flag is set in `psr`.

I think the corresponding source code can be found [here in lines 308-324](https://sourceware.org/git/?p=newlib-cygwin.git;a=blob;f=newlib/libc/stdlib/nano-mallocr.c;h=a2b50facc35ebdfb4fc3c8fb11cd170d634649b3;hb=HEAD#l308). The crash happens on dreferencing `r` in line 323.

```c
308        r = sbrk_aligned(RCALL alloc_size);
309
310        /* sbrk returns -1 if fail to allocate */
311        if (r == (void *)-1)
312        {
313            /* sbrk didn't have the requested amount. Let's check
314             * if the last item in the free list is adjacent to the
315             * current heap end (sbrk(0)). In that case, only ask
316             * for the difference in size and merge them */
317            p = free_list;
318            r = p;
319
320            while (r)
321            {
322                p=r;
323                r=r->next;
324            }
```
So the question is how can the processor enter the while loop if `r` is NULL?

**Comment by gschorcht** [comment]
It does indeed seem to be a bug in the `newlib` bundled with `gcc 10.3 2021.10`. The code in `gcc 12.2` seems to be slightly differen at the point in question:
```asm
   8001f4a:	f7ff ffbf 	bl	8001ecc <sbrk_aligned>
   8001f4e:	1c43      	adds	r3, r0, #1
   8001f50:	4604      	mov	r4, r0
   8001f52:	d158      	bne.n	8002006 <_malloc_r+0xf6>
   8001f54:	f8d8 4000 	ldr.w	r4, [r8]
   8001f58:	4627      	mov	r7, r4
   8001f5a:	2f00      	cmp	r7, #0
   8001f5c:	d143      	bne.n	8001fe6 <_malloc_r+0xd6>
-> 8001f5e:	2c00      	cmp	r4, #0
-> 8001f60:	d04b      	beq.n	8001ffa <_malloc_r+0xea>
   8001f62:	6823      	ldr	r3, [r4, #0]
```
It now tests `r4` for `0x0` before the instruction `ldr r3, [r4, #0]`.

**Comment by gschorcht** [comment]
This was obviously the [fix](https://sourceware.org/git/?p=newlib-cygwin.git;a=blobdiff;f=newlib/libc/stdlib/nano-mallocr.c;h=0c5fb2885278085cdeda2ae16d15799873cf061c;hp=6fb08a61607358e4897937892b46d9d8edacb844;hb=73d515fcfecd1004ccffef8fcd0c7223b9eec6ab;hpb=43ab30fbd50cd6394b6b6950647ce2c243fa80c5)

**Comment by gschorcht** [comment]
So we can't do anything from our side other than to upgrade to the `gcc 12.2 toolchain` which seems not to be the main stream if you open the official archive at https://developer.arm.com/downloads/-/gnu-rm.

**Comment by benpicco** [comment]
Fortunately that's also what Ubuntu ships these days

> `arm-none-eabi-gcc (15:12.2.rel1-1) 12.2.1 20221205`

We'll just have to update riotdocker

**Pull request merged**